File: implementacion_java/src/algoritmos/SequentialBlockIII.java
package algoritmos;

import utilidades.AlgoritmoMultiplicacion;
import utilidades.Convertidor;

import java.math.BigInteger;

public class SequentialBlockIII implements AlgoritmoMultiplicacion {

    Convertidor convertidor = new Convertidor();

    @Override
    public void multiplicar(BigInteger[][] matriz1, BigInteger[][] matriz2) {
        // Convertir de BigInteger a int
        int[][] doubleMatriz1 = convertidor.convertBigIntegerArrayToIntArray(matriz1);
        int[][] doubleMatriz2 = convertidor.convertBigIntegerArrayToIntArray(matriz2);
        int[][] doubleResultado = sequentialBlockIII3(doubleMatriz1, doubleMatriz2, doubleMatriz1.length);

        // Convertir el resultado int a BigINteger
        convertidor.convertIntArrayToBigInteger(doubleResultado);
    }

    // Método para multiplicar dos matrices en bloques con un tamaño de bloque específico
    public int[][] sequentialBlockIII3(int[][] A, int[][] B, int block_size) {
        int n = A.length;
        int[][] C = new int[n][n]; // Matriz para almacenar el resultado

        // Recorre las matrices en bloques de tamaño block_size
        for (int i = 0; i < n; i += block_size) {
            for (int j = 0; j < n; j += block_size) {
                for (int k = 0; k < n; k += block_size) {
                    // Multiplica los bloques de matrices
                    for (int ii = i; ii < i + block_size && ii < n; ii++) {
                        for (int jj = j; jj < j + block_size && jj < n; jj++) {
                            for (int kk = k; kk < k + block_size && kk < n; kk++) {
                                C[ii][jj] += A[ii][kk] * B[kk][jj]; // Actualiza la celda correspondiente en C
                            }
                        }
                    }
                }
            }
        }
        return C; // Devuelve el resultado de la multiplicación de matrices
    }

}

File: implementacion_kotlin/src/multiplicacion_matrices/SequentialBlockIII.kt
package multiplicacion_matrices

import utilidades.AlgoritmoMultiplicacion
import utilidades.Convertidor
import java.math.BigInteger

class SequentialBlockIII : AlgoritmoMultiplicacion {
    var convertidor: Convertidor = Convertidor()

    override fun multiplicar(matriz1: Array<Array<BigInteger>>, matriz2: Array<Array<BigInteger>>) {
        // Convertir de BigInteger a int
        val doubleMatriz1 = convertidor.convertBigIntegerArrayToIntArray(matriz1)
        val doubleMatriz2 = convertidor.convertBigIntegerArrayToIntArray(matriz2)
        val doubleResultado = sequentialBlockIII3(doubleMatriz1, doubleMatriz2, doubleMatriz1.size)

        // Convertir el resultado int a BigINteger
        convertidor.convertIntArrayToBigInteger(doubleResultado)
    }

    /**
     * @param A          matriz
     * @param B          matriz
     * @param block_size tamaño del bloque
     * @return matriz resultante de la multiplicación
     */
    fun sequentialBlockIII3(A: Array<IntArray>, B: Array<IntArray>, block_size: Int): Array<IntArray> {
        val n = A.size
        val C = Array(n) { IntArray(n) }

        // Recorre las matrices en bloques de tamaño block_size
        var i = 0
        while (i < n) {
            var j = 0
            while (j < n) {
                var k = 0
                while (k < n) {
                    // Multiplica los bloques de matrices
                    var ii = i
                    while (ii < i + block_size && ii < n) {
                        var jj = j
                        while (jj < j + block_size && ii < n) {
                            var kk = k
                            while (kk < k + block_size && k < n) {
                                C[ii][jj] += A[ii][kk] * B[kk][jj]

                                kk++
                            }
                            jj++
                        }
                        ii++
                    }
                    k += block_size
                }
                j += block_size
            }
            i += block_size
        }
        return C
    }
}
